src/Aspire.Dashboard/Components/Pages/TraceDetail.razor.cs

```diff
@@ -93,6 +94,43 @@ private ValueTask<GridItemsProviderResult<SpanWaterfallViewModel>> GetData(GridI
             Items = page.ToList(),
             TotalItemCount = visibleSpanWaterfallViewModels.Count
         });
+
+        // Return spans where any child span matches the filter.
+        List<SpanWaterfallViewModel> GetFilteredSpanWaterfalls(List<SpanWaterfallViewModel> viewModels)
+        {
+            var viewModelsToFilterMatch = new Dictionary<SpanWaterfallViewModel, bool>();
+            return [.. viewModels.Where(AnyChildMatchFilter)];
+
+            bool AnyChildMatchFilter(SpanWaterfallViewModel span)
+            {
+                if (viewModelsToFilterMatch.TryGetValue(span, out var matchesFilter))
+                {
+                    return matchesFilter;
+                }
+
+                matchesFilter = span.Children.Any(AnyChildMatchFilter) || Filter(_filter, span);
+                viewModelsToFilterMatch.Add(span, matchesFilter);
```

Comment:
Be defensive against a span someone getting into the list multiple times to avoid a "key already exists" error.
```suggestion
                viewModelsToFilterMatch.Add[span] = matchesFilter;
```